### PR TITLE
CHET-142: implement different modes for state machine enforcement

### PR DESCRIPTION
**Pull request by bpar476**
This is a proof of concept for how we could have multiple modes for the migration assistant.

Default - enforce the current stage is as expected before running the operation
no-verify - always run the migration operation regardless of the current stage. This lets you test the database migration stage without having to wait to have moved through a file system migration.
passthrough - doesn't run the operation and sets the stage to the specified short circuit stage. This is good for if you just want to click through the UI.

I'm not completely happy with this approach as it warps how you write tests for migration operations in a couple of ways:

* Your tests need to have a ModalMigrationStageWorker
* When you create a new operation as a part of the migration you need to remember to call the worker so that it does the mode check.

I wanted to use a template method approach at first however there would need to be an abstract class doing this logic for each migration operation so there is still a bit of pain every time you implement a new migration operation.

### CODE REVIEW

File: src/main/java/com/atlassian/migration/datacenter/core/ModalMigrationStageWorker.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        if (mode == null || mode.equals(DCMigrationAssistantMode.DEFAULT)) {
+            if (migrationService.getCurrentStage().equals(expectedCurrentStage)) {
+                operation.migrate();
+            }
+            return;
+        }
+
+        if (mode.equals(DCMigrationAssistantMode.PASSTHROUGH)) {
+            try {
```
Comment (bpar476) [inline src/main/java/com/atlassian/migration/datacenter/core/ModalMigrationStageWorker.java]: If an operation is meant to have side-effects we may need to change this from a simple transition to another operation which fills the migration with dummy data. We are already missing the application stack ID in the migration context after this.

**Comment by bpar476** [comment]
This PR is a draft until I get feedback on the approach.

**Comment by theghostwhoforks** [comment]
Now that I've seen a bit more of `AwsMigrationService`, I wonder if it makes sense to harvest the work done in this PR and inject the behaviour into the service based on a spring profile, say?

If we can apply an advice if a profile is set that will ignore the error that transition will raise, wouldn't that, in effect, give us the same effect as using the worker as described above?

Alternately, Getting the advice injected dynamically can be done by refreshing the bean based on a new (or edit of) plugin property

**Comment by bpar476** [comment]
> Now that I've seen a bit more of `AwsMigrationService`, I wonder if it makes sense to harvest the work done in this PR and inject the behaviour into the service based on a spring profile, say?
> 
> If we can apply an advice if a profile is set that will ignore the error that transition will raise, wouldn't that, in effect, give us the same effect as using the worker as described above?
> 
> Alternately, Getting the advice injected dynamically can be done by refreshing the bean based on a new (or edit of) plugin property

I'm all for implementing an alternative solution. This one isn't very Java-y. However, being able to set the mode through the UI when in development seems important to me. I'm not sure how you do this with Spring as I'm not deeply familiar. Keen to see what you mean

File: src/main/java/com/atlassian/migration/datacenter/core/ModalMigrationStageWorker.java (hunk 1 above)
Comment (theghostwhoforks) [reply src/main/java/com/atlassian/migration/datacenter/core/ModalMigrationStageWorker.java]: > > Now that I've seen a bit more of `AwsMigrationService`, I wonder if it makes sense to harvest the work done in this PR and inject the behaviour into the service based on a spring profile, say?
> > If we can apply an advice if a profile is set that will ignore the error that transition will raise, wouldn't that, in effect, give us the same effect as using the worker as described above?
> > Alternately, Getting the advice injected dynamically can be done by refreshing the bean based on a new (or edit of) plugin property
> 
> I'm all for implementing an alternative solution. This one isn't very Java-y. However, being able to set the mode through the UI when in development seems important to me. I'm not sure how you do this with Spring as I'm not deeply familiar. Keen to see what you mean

Do we need to expose this to customers via the UI or is this only for us (devs) to help with testing

**Comment by nanux** [comment]
Following up our IRL conversation I think the solution is trying to cover two cases.
# Add ability to jump into a stage that we need to be to test certain stages
# Add NoOp transition state machine that can be useful to go through when working on frontend.

I think at the moment we need the 1. and we should consider doing 2. on the frontend as this change has a quite big impact on production code to provide utility testing functionality. I would suggest introducing `setStage()` endpoint that would set a stage and would be exposed in developer mode and consider adding the dummy transition machine when we start working on frontend and potentially stub it in Typescript instead of backend.

**Pull request closed without merging**